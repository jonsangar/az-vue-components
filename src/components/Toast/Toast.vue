<template lang="pug">
transition(
  :name="transition"
  v-on:after-enter="onAfterEnter"
  v-on:after-leave="onAfterLeave")

  .toast(v-show="value" @click="onClick" :class="classMode")
    .toast__wrapper()
      .toast__body
        az-loading.toast__loading(type="circle" v-if="isLoading")
        az-icon.toast__icon(
          v-else-if="type === 'success'"
          :name="icon" scale="5")

        az-icon.toast__icon(
          v-else
          :name="icon" scale="3")

        //- .toast__icon(v-else-if="type !== 'warning' && type !== 'info'")
        //-   az-icon(:name="icon" scale="1.5")
        //- .toast__icon--warning(v-else)
        //-   az-icon(:name="icon" scale="3")

        .toast__title(
          :class="{'toast__title--onlyTitle': !message}") {{ title }}
        .toast__message(v-if="message") {{ message }}
        .toast__errorCode(v-if="errorCode && !isProd")
          | {{ errorCode }}
          span(v-if="errorName && !isProd") - {{ errorName }}
        .toast__errorDescription(v-if="errorDescription && !isProd") {{ errorDescription }}
        .toast__errorInfo(v-if="errorInfo && !isProd")
          strong {{ errorInfo.param }}:
          span.margin-left-quarter {{ errorInfo.value }}
</template>

<script>
export default {
  props: {
    icon: {
      type: String,
      default: '',
    },
    title: {
      type: String,
      default: '',
    },
    message: {
      type: String,
      default: '',
    },
    errorCode: {
      type: String,
      required: false,
    },
    errorName: {
      type: String,
      required: false,
    },
    errorDescription: {
      type: String,
      required: false,
    },
    errorInfo: {
      type: Object,
      required: false,
    },
    transition: {
      type: String,
      default: 'fade',
    },
    type: {
      type: String,
      default: 'text',
    },
    closeOnClick: {
      type: Boolean,
      default: true,
    },
    value: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    classMode() {
      if (this.type === 'loading') return 'toast--loading';
      if (this.type === 'success') return 'toast--success';
      if (this.type === 'warning') return 'toast--warning';
      if (this.type === 'error') return 'toast--danger';
      return '';
    },

    isLoading() {
      return this.type === 'loading';
    },

    isError() {
      return this.type === 'error';
    },

    isProd() {
      return this.$constants.IS_PRODUCTION;
    },
  },

  methods: {
    onClick() {
      if (this.closeOnClick) {
        this.clear();
      }
    },

    onAfterEnter() {
      this.$emit('opened');

      if (this.onOpened) {
        this.onOpened();
      }
    },

    onAfterLeave() {
      this.$emit('closed');
    },
  },

  mounted() {
    document.body.classList.add('toast-overflow');

    //
    // AÃ±ado el componente
    //
    document.body.appendChild(this.$el);
  },

  beforeDestroy() {
    document.body.classList.remove('toast-overflow');

    //
    // Elimino el componente
    //
    const parent = this.$el.parentNode;
    if (parent) parent.removeChild(this.$el);
  },
};
</script>

<style lang="scss" scoped>
.toast {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 100000;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100vw;
  height: 100vh;
  background-color: rgba(0, 0, 0, 0.8);

  &--loading {
    .toast__wrapper {
      background-image: none;
    }

    .toast__title {
      animation: flickerAnimation 2s infinite;
    }
  }

  &--warning {
    .toast__wrapper {
      background-image: radial-gradient(circle at 0% 0%, hsl(31, 100%, 60%), hsl(29, 100%, 35%));
    }
  }

  &--success {
    .toast__wrapper {
      background-image: radial-gradient(circle at 0% 0%, #00AAA1, #137D79);
    }
  }

  &--danger {
    .toast__wrapper {
      background-image: radial-gradient(circle at 0% 0%, hsl(9, 86%, 44%), hsl(5, 94%, 28%));
    }
  }

  &__wrapper {
    @include shadow(2);

    position: relative;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    width: 90%;
    max-width: 350px;
    padding: triple() simple() simple();
    color: white;
    text-align: center;
    background-image: radial-gradient(circle at 0% 0%, color(brand), color(brand-darker));
    border-radius: $border-radius;
    transition: all 0.3s;
  }

  &__support,
  &__close {
    position: absolute;
    top: simple();
    right: half();
    width: double();
    height: double();
    cursor: pointer;
  }

  &__body {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
  }

  &__icon {
    margin-bottom: double();
  }

  // &__icon {
  //   width: 4rem;
  //   height: 4rem;
  //   padding: half();
  //   margin-bottom: double();
  //   line-height: 2rem;
  //   text-align: center;
  //   border: 2px solid white;
  //   border-radius: 50%;
  // }

  // &__icon--warning {
  //   margin-bottom: double();
  // }

  &__loading {
    margin-bottom: simple();
    font-size: 2rem !important;
  }

  &__title {
    margin-bottom: half();
    font-size: font-size(h4);
    font-weight: font-weight(bold);

    &--onlyTitle {
      font-size: font-size(h5);
    }
  }

  &__errorCode {
    margin-top: simple();
    font-weight: font-weight(bold);
  }

  &__errorInfo {
    margin-top: half();
    font-size: font-size(small1);
  }

  &__footer {
    width: 100%;
    margin-top: double();
  }

  &__buttons {
    display: flex;
    width: 90%;
    margin: 0 auto half();
  }

  &__button {
    width: 100%;
    height: double();
    margin-right: simple();
    font-weight: font-weight(bold);
    line-height: double();
    cursor: pointer;
    border: solid 1px white;
    border-radius: $border-radius;

    &:last-of-type {
      margin-right: 0;
    }
  }

  &__link {
    text-decoration: underline;
    cursor: pointer;
  }

  ::v-deep {
    .az-loading__item::before {
      background-color: color(gray6);
    }

    .az-loading__path {
      stroke: color(gray7);
    }
  }
}

@keyframes flickerAnimation {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.7;
  }
  100% {
    opacity: 1;
  }
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s;
}
.fade-enter, .fade-leave-to /* .fade-leave-active below version 2.1.8 */ {
  opacity: 0;
}
</style>
