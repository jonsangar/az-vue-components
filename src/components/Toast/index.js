import Vue from 'vue';
import i18n from 'Core/services/language.service';
import VueToast from './Toast.vue';

const defaultOptions = {
  closeOnClick: true,
  duration: 3000,
  icon: '',
  title: '',
  message: '',
  errorCode: '',
  errorName: '',
  errorDescription: '',
  errorInfo: null,
  transition: 'fade',
  type: 'loading',
  onClose: null,
  onOpened: null,
  value: true,
};

let toastInstance = null;

/**
 * El componente toast permite un uso simplificado donde unicamente pasamos el mensaje.
 * Cuando se pasa un mensaje como opción, tenemos que parsearlo como un objeto
 *
 * @param {Object,String} title
 */
function parseOptions(title) {
  if (typeof title === 'string') {
    return { title };
  }

  return title;
}

/**
 * Crea una nueva instancia del componente,
 * si ya existe una devuelve la que había
 */
function createInstance() {
  if (toastInstance) return toastInstance;

  toastInstance = new (Vue.extend(VueToast))({
    el: document.createElement('div'),
  });

  return toastInstance;
}

/**
 * Crea un Objeto Toast
 *
 * @param {Object} options - Opciones del Toast
 */
function createToast(options) {
  const toast = createInstance();

  options = { ...defaultOptions, ...options };

  //
  // Le añado un método para eliminarlo
  //    - Si tiene definido un evento onClose lo ejecuto
  //    - Limpio el timeout para cerrarse automaticamente
  //    - Elimino el nodo del DOM
  //    - Elimino el componente
  //
  options.clear = () => {
    toast.value = false;

    if (options.onClose) {
      options.onClose();
    }

    clearTimeout(toast.timer);
    toastInstance = null;

    const parent = toast.$el.parentNode;
    if (parent) parent.removeChild(toast.$el);

    toast.$destroy();
  };

  //
  // Le paso las opciones a la instancia del toast
  //
  Object.assign(toast, options);

  //
  // Le añado un timeout para que se oculte automaticamente
  //
  clearTimeout(toast.timer);

  if (options.duration > 0) {
    toast.timer = setTimeout(() => {
      toast.clear();
    }, options.duration);
  }

  return toast;
}

/**
 * Creo el objeto Toast que contiene los métodos para crear los diferentes tipos de toast
 */
const Toast = {
  loading: options =>
    createToast({
      type: 'loading',
      icon: 'loading',
      title: i18n.t('messages.loading'),
      duration: 0,
      closeOnClick: false,
      ...parseOptions(options),
    }),

  info: options =>
    createToast({
      type: 'info',
      icon: 'info',
      ...parseOptions(options),
    }),

  success: options =>
    createToast({
      type: 'success',
      icon: 'check',
      title: i18n.t('messages.success'),
      ...parseOptions(options),
    }),

  fail: options =>
    createToast({
      type: 'warning',
      icon: 'warning',
      title: i18n.t('messages.errors.defaultError.title'),
      message: i18n.t('messages.errors.defaultError.message'),
      ...parseOptions(options),
    }),

  error: error => {
    const title = `${error.module}.${error.name}.title`;
    const message = `${error.module}.${error.name}.message`;

    const titleString = i18n.te(title) ? i18n.t(title) : i18n.t('messages.errors.defaultError.title');
    const messageString = i18n.te(message) ? i18n.t(message) : i18n.t('messages.errors.defaultError.message');

    createToast({
      type: error.isBlocking ? 'error' : 'warning',
      icon: error.icon || 'warning',
      title: titleString,
      message: messageString,
      errorCode: error.code || '',
      errorName: error.name || '',
      errorDescription: error.description || '',
      errorInfo: error.info || null,
      support: error.support || '',
      ...parseOptions(error),
    });
  },
};

/**
 * Le añado un método para eliminar el toast
 */
Toast.clear = () => {
  if(toastInstance) toastInstance.clear();
}

Toast.install = () => {
  Vue.use(VueToast);
};

Vue.prototype.$toast = Toast;

export default Toast;
