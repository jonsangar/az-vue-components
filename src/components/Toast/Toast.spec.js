import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Toast from './Toast.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Toats.vue', () => {

  const build = () => {
    const wrapper = shallowMount(Toast, {
      localVue,
      props: {
        value: true,
        type: 'error',
        icon: 'warning',
        title: 'Error'
      },
      mocks: {
        $constants: {
          IS_PRODUCTION: false
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Toast con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Comprueba que emite el evento opened con el evento after-enter', async () => {
    const { wrapper } = build()

    await wrapper.vm.onAfterEnter()
    expect(wrapper.emitted().opened).toBeTruthy()
    expect(wrapper.emitted().opened.length).toBe(1)
    expect(wrapper.emitted().opened[0]).toEqual([])
  });

  it('Comprueba que emite el evento closed con el evento after-leave', async () => {
    const { wrapper } = build()

    await wrapper.vm.onAfterLeave()
    expect(wrapper.emitted().closed).toBeTruthy()
    expect(wrapper.emitted().closed.length).toBe(1)
    expect(wrapper.emitted().closed[0]).toEqual([])
  });

  it('Comprueba que se muestra el título correctamente cuando el Toast tiene título y es de tipo error', async () => {
    const { wrapper } = build()

    expect(wrapper.find('.toast__title--onlyTitle').exists()).toBe(true)
  });
})

describe('Toats.vue', () => {

  const build = () => {
    const wrapper = shallowMount(Toast, {
      localVue,
      propsData: {
        value: true,
        type: 'error',
        icon: 'warning',
        title: 'Error',
        message: 'Error'
      },
      mocks: {
        $constants: {
          IS_PRODUCTION: false
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que se muestra el mensaje correctamente cuando el Toast tiene mensaje y es de tipo error', async () => {
    const { wrapper } = build()

    expect(wrapper.find('.toast__message').exists()).toBe(true)
  });
})

describe('Toats.vue', () => {

  const build = () => {
    const wrapper = shallowMount(Toast, {
      localVue,
      propsData: {
        value: true,
        type: 'error',
        icon: 'warning',
        errorCode: 'Error'
      },
      mocks: {
        $constants: {
          IS_PRODUCTION: false
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que se muestra el código de error correctamente cuando el Toast tiene código de error y es de tipo error', async () => {
    const { wrapper } = build()

    expect(wrapper.find('.toast__errorCode').exists()).toBe(true)
  });
})

describe('Toats.vue', () => {

  const build = () => {
    const wrapper = shallowMount(Toast, {
      localVue,
      propsData: {
        value: true,
        type: 'error',
        icon: 'warning',
        errorInfo: { param: 'Error', value: '0x01' }
      },
      mocks: {
        $constants: {
          IS_PRODUCTION: false
        }
      }
    })

    return {
      wrapper,
    }
  }

  it('Comprueba que se muestra la información del error correctamente cuando el Toast tiene información del error y es de tipo error', async () => {
    const { wrapper } = build()

    expect(wrapper.find('.toast__errorInfo').exists()).toBe(true)
  });
});
