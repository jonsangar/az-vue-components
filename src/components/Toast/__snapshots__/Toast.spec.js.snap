// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toats.vue Comprueba renderizado del componente Toast con snapshot 1`] = `
<transition-stub name="fade">
  <div class="toast" style="display: none;">
    <div class="toast__wrapper">
      <div class="toast__body">
        <az-icon-stub width="" height="" scale="3" fill="true" class="toast__icon"></az-icon-stub>
        <div class="toast__title toast__title--onlyTitle"></div>
        <!---->
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</transition-stub>
`;
